source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,360 @@
+                                          const bool two) {
+  CHECK_LENGTH(len, 14);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+  bool watch;
+  BufferHelper::peekBool(data, offset, watch);
+
+  callbacks_.onGetChildrenRequest(path, watch, two);
+}
+
+void DecoderImpl::parseDeleteRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 16);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+  int32_t version;
+  BufferHelper::peekInt32(data, offset, version);
+
+  callbacks_.onDeleteRequest(path, version);
+}
+
+void DecoderImpl::parseExistsRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 13);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+  bool watch;
+  BufferHelper::peekBool(data, offset, watch);
+
+  callbacks_.onExistsRequest(path, watch);
+}
+
+void DecoderImpl::parseGetAclRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 8);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+
+  callbacks_.onGetAclRequest(path);
+}
+
+void DecoderImpl::parseSetAclRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 8);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+  skipAcls(data, offset);
+  int32_t version;
+  BufferHelper::peekInt32(data, offset, version);
+
+  callbacks_.onSetAclRequest(path, version);
+}
+
+void DecoderImpl::parseSyncRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 8);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+
+  callbacks_.onSyncRequest(path);
+}
+
+void DecoderImpl::parseCheckRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 8);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+  int32_t version;
+  BufferHelper::peekInt32(data, offset, version);
+
+  callbacks_.onCheckRequest(path, version);
+}
+
+void DecoderImpl::parseMultiRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  // Treat empty transactions as a decoding error, there should be at least 1 header.
+  CHECK_LENGTH(len, 17);
+
+  while (true) {
+    int32_t type;
+    BufferHelper::peekInt32(data, offset, type);
+    bool done{};
+    BufferHelper::peekBool(data, offset, done);
+    int32_t error;
+    BufferHelper::peekInt32(data, offset, error);
+
+    if (done) {
+      break;
+    }
+
+    switch (type) {
+    case enumToInt(OpCodes::CREATE):
+      parseCreateRequest(data, offset, len, false);
+      break;
+    case enumToInt(OpCodes::SETDATA):
+      parseSetRequest(data, offset, len);
+      break;
+    case enumToInt(OpCodes::CHECK):
+      parseCheckRequest(data, offset, len);
+      break;
+    default:
+      // Should not happen.
+      throw new EnvoyException("Unknown type within a transaction");
```

Comment:
This leaks memory since you're trying a pointer, you want to throw a value. Just remove the `new`